File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        vector<vector<int>>ans;
        sort(nums.begin(),nums.end());
        int prev = INT_MIN;
        for(int i=0; i<nums.size();i++){
            
            if(nums[i] == prev) continue;
            prev = nums[i];
            
            int left = i+1;
            int right = nums.size()-1;
            
            int n = -1 * nums[i];
            
            while(left< right){
                
                int sum = nums[left] + nums[right];
                if(sum == n){
                    ans.push_back({nums[left],nums[right],nums[i]});
                    left++;
                    right--;
                    while(left < right and nums[left-1] == nums[left]) left++;
                    while(left < right and nums[right+1] == nums[right]) right--;
                } 
                
                if(sum < n) left++;
                else if(sum>n) right--;
                    
            }
            
        }
        
        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        nums.sort()
        ans =[]
        
        for i in range(len(nums)):
            
            if i>0 and nums[i] == nums[i-1]: continue
            
            left = i+1
            right = len(nums)-1
            target = -nums[i]
            
            while left < right:
                
                if nums[left] + nums[right] > target: right -=1
                elif nums[left] + nums[right]< target: left +=1
                else: 
                    ans.append([nums[i],nums[left],nums[right]])
                    left +=1
                    right -=1
                    
                    while left < right and nums[left] ==  nums[left-1]: left +=1
                    while left < right and nums[right] == nums[right+1]: right -=1
            
        return ans
                 
                    
                
                
                
                
                
                
        